swarm/storage/memstore_lrucache.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+// memory storage layer for the package blockhash
+
+package storage
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+
+	lru "github.com/hashicorp/golang-lru"
+)
+
+type MemStore struct {
+	cache    *lru.Cache
+	requests *lru.Cache
+	mu       sync.Mutex
+	disabled bool
+}
+
+func NewMemStore(_ *LDBStore, cacheCapacity uint, requestsCapacity uint) (m *MemStore) {
+	if cacheCapacity == 0 {
+		return &MemStore{
+			disabled: true,
+		}
+	}
+
+	onEvicted := func(key interface{}, value interface{}) {
+		v := value.(*Chunk)
+		<-v.dbStoredC
+	}
+	c, err := lru.NewWithEvict(int(cacheCapacity), onEvicted)
+	if err != nil {
+		panic(err)
+	}
+
+	r, err := lru.New(int(requestsCapacity))
+	if err != nil {
+		panic(err)
+	}
+
+	return &MemStore{
+		cache:    c,
+		requests: r,
+	}
+}
+
+func (m *MemStore) Get(key Key) (*Chunk, error) {
+	if m.disabled {
+		return nil, ErrChunkNotFound
+	}
+
+	m.mu.Lock()
+	defer m.mu.Unlock()
```

Comment:
Although the Get function of the LRU Cache can also change the contents of the cache, it uses a Lock itself, it is thread safe. So in our code I think it is enough to RLock here. What could that cause? Parallely running requests.Get and cache.Get (because they are different cache objects), but I just don't see what problems could that cause.

Reply:
Not sure what you are proposing here? I am also not fully confident about the relationship between the two caches.

Reply:
I propose just to RLock, and not Lock in the Get function